root/hist/hist/THistFillBenchmarks.cxx

```diff
@@ -16,69 +13,227 @@ static void FillArguments(benchmark::internal::Benchmark *b)
          b->Args({i, j, j / 2});
 }
 
+// Create array for x coordinates
+double * MakeXCoorVec(int nbOfDataPoints) 
+{
+   double * vec = new double[nbOfDataPoints];
+   for (int j = 0; j < nbOfDataPoints; ++j) {
+      double d = static_cast< double >(j);
+      vec[j] = d / nbOfDataPoints;
+   }
+   return vec;
+}
 
+// Create array for y coordinates
+double * MakeYCoorVec(int nbOfDataPoints) 
+{
+   double * vec = new double[nbOfDataPoints];
+   for (int j = 0; j < nbOfDataPoints; ++j) {
+      double d = static_cast< double >(j);
+      vec[j] = d / (nbOfDataPoints / 10);
+   }
+   return vec;
+}
+
+// Create array for weights
+double * MakeWeightVec(int nbOfDataPoints) 
+{
+   double * vec = new double[nbOfDataPoints];
+   for (int j = 0; j < nbOfDataPoints; ++j) {
+      double d = static_cast< double >(j);
+      vec[j] = d;
+   }
+   return vec;
+}
+
+// Arrays of coordinates and weights
+double * xCoords = MakeXCoorVec(3000000);
+double * yCoords = MakeYCoorVec(3000000);
+double * weights = MakeWeightVec(3000000);
+
+// Histograms with different bin configurations for Fill
+TH2D hist1{"hist1", "", 10, 0., 1., 10 / 2, 0., 10.};
+TH2D hist2{"hist2", "", 10, 0., 1., 100 / 2, 0., 10.};
+TH2D hist3{"hist3", "", 10, 0., 1., 1000 / 2, 0., 10.};
+TH2D hist4{"hist4", "", 100, 0., 1., 10 / 2, 0., 10.};
+TH2D hist5{"hist5", "", 100, 0., 1., 100 / 2, 0., 10.};
+TH2D hist6{"hist6", "", 100, 0., 1., 1000 / 2, 0., 10.};
+TH2D hist7{"hist7", "", 1000, 0., 1., 10 / 2, 0., 10.};
+TH2D hist8{"hist8", "", 1000, 0., 1., 100 / 2, 0., 10.};
+TH2D hist9{"hist9", "", 1000, 0., 1., 1000 / 2, 0., 10.};
 
 // Benchmark for simple Fill on 2D hist
 static void BM_THist_Fill(benchmark::State &state)
 {
+   int nbOfDataPoints = state.range(0); 
    int nbOfBinsForFirstAxis = state.range(1);
    int nbOfBinsForSecondAxis = state.range(2);
 
-   TH2D hist{"hist", "", nbOfBinsForFirstAxis, 0., 1., nbOfBinsForSecondAxis, 0., 10.};
-
-   int nbOfDataPoints = state.range(0); 
    for (auto _ : state) {
-      // Fill without weight
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10));
-      }
-
-      // Fill with weights
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10), d);
+      if (nbOfBinsForFirstAxis == 10) {
```

Comment:
All of those are `TH2D`s, so we could have a `map<int, map<int, TH2D*>>` that allows you to do `hist[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis]->Fill...`? And then move the two `for (...nbOfDataPoints...)` loops into a separate function, such that you just call `Fill(hist[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis])`? That would save many many lines of repeated code.